### PR TITLE
Feat: Elasticsearch APIs, BatchQueries + Unit tests

### PR DESCRIPTION
**Pull request by dikshamodi**

### CODE REVIEW

File: src/main/java/com/hyperlocal/server/ShopController.java [hunk 1]
```diff
@@ -1,64 +1,233 @@
...
+        .thenApply((responseString) -> {
+          JsonObject obj = JsonParser.parseString(responseString).getAsJsonObject();
+          JsonArray idListJson = obj.get("hits").getAsJsonObject().get("hits").getAsJsonArray();
+          for (JsonElement id : idListJson) {
+            shopIDList.add(id.getAsJsonObject().get("_id").getAsLong());
+          }
+          return shopIDList;
+        
+        // Perform BatchQuery on shopIDs and get List of ShopDetails corresponding to the shopIDs
+        }).thenCompose((shopList) -> {
```
Comment (onish-garg) [inline src/main/java/com/hyperlocal/server/ShopController.java:118]: In lieu of chaining a `thenCompose()` after the `thenApply()` method, you can get rid of `thenApply()` method altogether and call `getShopsByShopIDBatch()` in the same lambda function which prepares the `shopIDList`.

Comment (dikshamodi) [reply src/main/java/com/hyperlocal/server/ShopController.java:118]: Actually. Thanks for this!! Let me make the change at other places too

File: src/main/java/com/hyperlocal/server/ShopController.java [hunk 2]
```diff
@@ -1,64 +1,233 @@
...
+        .thenApply((responseString) -> {
+          JsonObject obj = JsonParser.parseString(responseString).getAsJsonObject();
+          JsonArray idListJson = obj.get("hits").getAsJsonObject().get("hits").getAsJsonArray();
+          for (JsonElement id : idListJson) {
+            shopIDList.add(id.getAsJsonObject().get("_id").getAsLong());
+          }
+          return shopIDList;
+      
+        // Get List of ShopDetails for the list of ShopIDs via Batch Query
+        }).thenCompose((shopList) -> {
```
Comment (onish-garg) [inline src/main/java/com/hyperlocal/server/ShopController.java:155]: In lieu of chaining a `thenCompose()` after the `thenApply()` method, you can get rid of `thenApply()` method altogether and call `getShopsByShopIDBatch()` in the same lambda function which prepares the `shopIDList`.

File: src/test/java/com/hyperlocal/server/ShopControllerTest.java [hunk 3]
```diff
@@ -64,20 +70,14 @@ public void shouldGetShopsByMerchantID() throws Exception {
     assertThat(controller).isNotNull();
     String merchantID = "1000000000000";
 
-    RowData shopRecord = new FakeRowData(
-      "ShopID", 1L,
```
Comment (onish-garg) [inline src/test/java/com/hyperlocal/server/ShopControllerTest.java:68]: Might be a good idea to preserve this indentation as it make the `ColumnName => ColumnValue` mapping readable

File: src/test/java/com/hyperlocal/server/ShopControllerTest.java [hunk 4]
```diff
@@ -156,43 +142,46 @@ public void shouldUpsertCatalog() throws Exception {
...
+    when(connection.sendQuery("COMMIT")).thenReturn(CompletableFuture.completedFuture(emptyQueryResult));
+    when(template.publish("projects/speedy-anthem-217710/topics/testTopic", "1000000000000"))
+        .thenReturn(new AsyncResult<>("DONE"));
 
     /* ACT */
-    CompletableFuture<HashMap<String, Object>> actualMapPromise = controller.upsertCatalog(shopID, new Gson().toJson(payload));
+    CompletableFuture<HashMap<String, Object>> actualMapPromise = controller.upsertCatalog(shopID,
+        new Gson().toJson(payload));
+
+    verify(template).publish("projects/speedy-anthem-217710/topics/testTopic", "1000000000000");
```
Comment (onish-garg) [inline src/test/java/com/hyperlocal/server/ShopControllerTest.java:184]: This verification can go under the `ASSERT` section starting at (new) line 186.

File: src/main/java/com/hyperlocal/server/ShopController.java (hunk 2 above)
Comment (dikshamodi) [reply src/main/java/com/hyperlocal/server/ShopController.java]: Done. Thanks a lot!!!

File: src/test/java/com/hyperlocal/server/ShopControllerTest.java (hunk 3 above)
Comment (dikshamodi) [reply src/test/java/com/hyperlocal/server/ShopControllerTest.java:68]: Done. Thanks a lot!!!

File: src/test/java/com/hyperlocal/server/ShopControllerTest.java (hunk 4 above)
Comment (dikshamodi) [reply src/test/java/com/hyperlocal/server/ShopControllerTest.java]: Done. Thanks a lot!!!

File: src/main/java/com/hyperlocal/server/Utilities.java [hunk 5]
```diff
@@ -17,4 +17,13 @@ public static boolean doubleThresholdCompare(Double numA, Double numB, Double th
...
         return Math.abs(numA-numB) < threshold;
     }
+
+    // Create a ?,?,? placeholder with numOfPlaceholders '?' to use in SQL prepared statements
+    public static String getPlaceHolderString(Integer numOfPlaceholders) {
+      StringBuilder result = new StringBuilder();
+      for (Integer i = 0; i < numOfPlaceholders; i++) {
+        result.append("?,");
+      }
+      return result.length() > 0 ? result.substring(0, result.length() - 1) : "";
```
Comment (onish-garg) [inline src/main/java/com/hyperlocal/server/Utilities.java:27]: I think it's fair to assume `numOfPlaceholders >= 1`. In that case:
```suggestion
      StringBuilder result = new StringBuilder("?");
      for (Integer i = 0; i < numOfPlaceholders-1; i++) {
        result.append(",?");
      }
      return result.toString();
```

File: src/main/java/com/hyperlocal/server/ShopController.java [hunk 6]
```diff
@@ -1,64 +1,227 @@
...
 import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.server.ResponseStatusException;
 
 @RestController
 public class ShopController {
 
-  private final PubSubTemplate publisher;
-  private static final String DATABASE_URL = "jdbc:mysql:///hyperlocal?socketFactory=com.google.cloud.sql.mysql.SocketFactory&cloudSqlInstance=speedy-anthem-217710:us-central1:hyperlocal";;
+  private PubSubTemplate publisher;
+  private static final String DATABASE_URL = "jdbc:mysql://10.124.32.3:3306/hyperlocal";
```
Comment (nagasrinivas-google) [inline src/main/java/com/hyperlocal/server/ShopController.java:54]: Are these static ips? Share some link on how to get to this ip address?

File: src/main/java/com/hyperlocal/server/ShopController.java [hunk 7]
```diff
@@ -1,64 +1,227 @@
...
+  private static final String SEARCH_INDEX_URL = "http://10.128.0.13:9200/shops/_search?filter_path=hits.hits._id";
```
Comment (nagasrinivas-google) [inline src/main/java/com/hyperlocal/server/ShopController.java:55]: Are these static ips? Share some link on how to get to this ip address?

File: src/test/java/com/hyperlocal/server/ShopControllerTest.java [hunk 8]
```diff
@@ -38,15 +41,19 @@
   private static final String SHOP_INSERT_STATEMENT = "INSERT INTO `Shops` (`ShopName`, `TypeOfService`, `Latitude`, `Longitude`, `AddressLine1`, `MerchantID`) VALUES (?,?,?,?,?,?);";;
```
Comment (nagasrinivas-google) [inline src/test/java/com/hyperlocal/server/ShopControllerTest.java:41]: Can you reuse Strings from Constants and avoid declarations in Tests

File: pom.xml [hunk 9]
```diff
@@ -61,6 +68,12 @@
       <artifactId>spring-cloud-gcp-starter-pubsub</artifactId>
     </dependency>
 
+    <dependency>
```
Comment (nagasrinivas-google) [inline pom.xml:71]: Btw, Is this necessary?

Comment (dikshamodi) [reply pom.xml:71]: I'm not sure. I'll check without it once

File: src/test/java/com/hyperlocal/server/ShopControllerTest.java (hunk 8 above)
Comment (dikshamodi) [reply src/test/java/com/hyperlocal/server/ShopControllerTest.java:41]: Working on it. Thanks a lot!!!

**Pull request merged**